### PR TITLE
293 Convert compare_outputs.py into an 'output_checkers' interface

### LINKED ISSUE #293: Create interface for "compare_outputs"
# Requested Update

### Description
* Currently, the "compare_outputs.py" module in the main geoips repo contains a number of different functions for comparing different types of outputs.
* This module should be turned into an interface, to allow modularly specifying the various output comparisons.
* This will also allow easily adding more types of output comparisons (ie, pure python based comparisons to remove reliance on imagemagick, as well as non-image based comparisons to reduce reliance on matplotlib/cartopy versions when comparing outputs).
* Currently compare_outputs is not set up as an interface, so there will be some effort required to rearrange it into a format conducive to using as an interface.

### Background and Motivation
*  [ OPTIONAL: Provide additional detailed information and background related to this issue ]

### Alternative Solutions
*  [ OPTIONAL: If appropriate, list alternative fixes that could be considered]
  
### Environment
*  [ IF APPROPRIATE: Corresponding repository versions ]

### Code to demonstrate issue
*  [ IF APPROPRIATE: Can include steps here to identify/demonstrate bug fix or desired feature ]

### Checklist for Completion
* [ ]

### PR DESCRIPTION
**Pull request by evrose54**
# Reviewer Checklist

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
  * [x] None required -- Fully tested already.
* [x] Required ***unit tests*** added and pass for new/modified functionality
  * [x] None required -- None needed for this branch. Suite of tests cover these updates.
* [x] Required ***integration tests*** added and pass for new/modified functionality
  * [x] None required. -- Current integration tests do the job fine.
* [x] Required ***documentation*** added for new/modified functionality
  * [x] None required -- Documentation has been added to the code.
* [x] Required ***release notes*** added for new/modified functionality
  * [x] None required -- Release notes added.

# Related Issues
fixes #293 

# Testing Instructions
To test this, run geoips/tests/integration_tests/full_test.sh. This PR is still a draft and is not ready for merge yet, however the tests are passing at the moment!

# Summary
Currently, GeoIPS compares output products with a python file called 'compare_outputs.py'. While this worked, we really should compare outputs using a GeoIPS interface. This PR has created an 'output_checkers' interface, which removes the need to use the compare outputs python script directly.

Files Modified:

  - geoips/interfaces/__init__.py
  - geoips/interfaces/base.py
  - geoips/plugins/procflows/single_source.py
  - geoips/plugins/procflows/config_based.py
  - pyproject.toml

Files Added:

  - geoips/interfaces/module_based/output_checkers.py
  - geoips/plugins/modules/output_checkers/geotiff_checker.py
  - geoips/plugins/modules/output_checkers/gz_checker.py
  - geoips/plugins/modules/output_checkers/image_checker.py
  - geoips/plugins/modules/output_checkers/netcdf_checker.py
  - geoips/plugins/modules/output_checkers/text_checker.py

Files Deleted:

  - geoips/compare_outputs.py

### CODE REVIEW

File: geoips/plugins/modules/procflows/single_source.py
Comment (mindyls) [inline geoips/plugins/modules/procflows/single_source.py:1382]: Ahh, so now it will find installed output checkers for indiviidual data types in any repo.  That is good.

File: pyproject.toml
Comment (mindyls) [inline pyproject.toml:3]: What linting are you using for TOML?  We should probably set that up in general, or we're going to end up with a lot of reformatting changes like this...

File: pyproject.toml
Comment (mindyls) [inline pyproject.toml:266]: This is the only actual change to pyproject.toml, correct?

**Review by mindyls — APPROVED** [review]
Merged #345 #351 #343 #330 into a single branch and ran full_test.sh on all of them:

I am good to merge once my conversations/questions are resolved.

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [inline geoips/interfaces/module_based/output_checkers.py:61]: Shouldn't these be automatically discovered, without explicitly importing each one?  We want to be able to add new output checkers to other repos, without referencing them here?

File: geoips/plugins/modules/output_checkers/geotiff_checker.py
Comment (mindyls) [inline geoips/plugins/modules/output_checkers/geotiff_checker.py:112]: Can this just call "geotiffs_match" directly, without going through the general "compare_outputs" module?  I would assume once we are inside a given output checker, we know what the format is/should be, and can just check it directly?

File: geoips/plugins/modules/output_checkers/utils/compare_outputs.py
Comment (mindyls) [inline geoips/plugins/modules/output_checkers/utils/compare_outputs.py:176]: I think this would all be better directly in the interface, so the actual output checkers are standalone.  Ie, from within the interface, atuomatically identify all of the output checkers, then find call "correct_type" for each one until you find the right one.

We should probably eventually have a way to explicitly specify the output checker type, but for now looping through all of them until we find the right one is probably fine (that's what it's doing now)

File: geoips/plugins/modules/procflows/config_based.py
Comment (mindyls) [inline geoips/plugins/modules/procflows/config_based.py:1698]: Maybe this should be "select_checker" or "identify_checker", to differentiate from the standard "get" interface method

Ie, like all the other interfaces, if you do "output_checkers.get("geotiff_checker") it will return the explicitly named geotiff_checker.  We can use that if we want to eventually specify which checker specifically we should use.

We can attach a helper function called 
``checker_plugin_name = output_checkers.identify_checker("<filename>")``
, which will return, say "geotiff_checker", then you can do
``checker_plugin.get(checker_plugin_name)``

File: geoips/plugins/modules/procflows/single_source.py
Comment (mindyls) [inline geoips/plugins/modules/procflows/single_source.py:1862]: See comment above.  This could either be "output_checkers.find_checker" and return the actual plugin, as you are doing here.  Or separate it out into "output_checkers.identify_checker" and output_checkers.get(<checker_name>)

**Review by mindyls — CHANGES_REQUESTED** [review]
I left a few comments on how I think we might want to restructure the interface slightly.  @jsolbrig  what do you think?

File: pyproject.toml
Comment (evrose54) [reply pyproject.toml:266]: Yup. Not sure why there are other things included in this diff. This is the only change I made to pyproject.toml

File: pyproject.toml
Comment (evrose54) [reply pyproject.toml:3]: How can I check that? I'm not sure.

File: geoips/plugins/modules/procflows/config_based.py
Comment (evrose54) [reply geoips/plugins/modules/procflows/config_based.py:1698]: That's not a bad idea. Or we can create just one function like "select_checker" as you mentioned above, which automatically gets the checker for us. Seems unnecessary to have it split up into two functions, however I'm happy to implement whichever you think is more appropriate. If we end up wanting to grab a checker via get, the two method approach might be better suited.

File: geoips/interfaces/module_based/output_checkers.py
Comment (evrose54) [reply geoips/interfaces/module_based/output_checkers.py:61]: You're right. That was a brain fart on my side! I'll change that now.

File: geoips/plugins/modules/procflows/config_based.py
Comment (mindyls) [reply geoips/plugins/modules/procflows/config_based.py:1698]: Either way is fine. output_checkers.get() is already a built in interface function that is on all interfaces, right @jsolbrig ?

File: pyproject.toml
Comment (mindyls) [reply pyproject.toml:3]: It is probably within your VSCode settings.  @jsolbrig  ?

File: geoips/plugins/modules/procflows/single_source.py
Comment (evrose54) [reply geoips/plugins/modules/procflows/single_source.py:1862]: Updated the function signature to be 'select_checker'

File: geoips/plugins/modules/output_checkers/geotiff_checker.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/geotiff_checker.py:112]: I think as long as this ALWAYS imports compare_outputs from the geoips repo that is probably fine (ie, there is no reference to the explicit output_checker plugins within compare_outputs - it uses get_plugins to implicitly find all the output checker plugins in any repo).  We just don't want to go back to how it was when recenter_tc and geoips repos both had completely separate "compare_outputs" Modules...

Comment (mindyls) [reply geoips/plugins/modules/output_checkers/geotiff_checker.py:112]: Actually, maybe the "compare_outputs" utility should be attached to the interface itself, rather than a separate python script in output_checkers/utils ?  @jsolbrig should probably chime in on this.

**Comment by evrose54** [comment]
I updated recenter_tc and data_fusion package to use the output_checkers interface

File: geoips/plugins/modules/output_checkers/geotiff_checker.py
Comment (evrose54) [reply geoips/plugins/modules/output_checkers/geotiff_checker.py:112]: I'll wait to update this and the comment below until we hear back from @jsolbrig. Might be a good idea to attach it to the interface, so I'll wait before I change anything in the 'compare_outputs' code.

File: pyproject.toml
Comment (jsolbrig) [reply pyproject.toml:3]: I think it's probably "Even Better TOML".

File: geoips/plugins/modules/output_checkers/geotiff_checker.py
Comment (jsolbrig) [reply geoips/plugins/modules/output_checkers/geotiff_checker.py]: As an update here, we're discussing how best to handle use of methods from inherited BasePlugin classes. Say we define `OutputCheckersBasePlugin.compare_outputs()`, how can that be accessed from within a plugin's `call()` function.

Currently, a plugin's `call()` function is attached to the plugin's class as a static method like this:
```
obj_attrs["__call__"] = staticmethod(getattr(plugin_module, "call"))
```
This makes is so that the resulting plugin object is callable and will execute `plugin_module.call()` when the object is called. So:
```
plg(foo, bar)
```
calls
```
plg.__call__(foo, bar)
```
which calls
```
plugin_module.call(foo, bar)
```

Since this is implemented as a static method, though, it does not provide access to `self` for the object. This means that we can't access any of the plugin object's instance methods. So, if we define `OutputCheckersBasePlugin.compare_outputs()` we can't access it from within `call()`.

There are a few methods to make instance methods available from within `call()`:

1. We could require all plugin modules' `call()` functions to include one additional argument such as `self` or 
    `plugin`, then attach `call()` to the plugin object as an instance method. using:
    ```
    obj_attrs["__call__"] = instancemethod(getattr(plugin_module, "call"))
    ```
    This would allow us to do something like this:
    ```
    def call(self, foo, bar):
        self.compare_outputs()
    ```
2. Rather than requiring `self` or `plugin` as arguments, we could make them optional, then attach the
    `call()` functions to their classes as staticmethods if `self` or `plugin` is absent and as instancemethods if 
    they are present.
3. We could create an additional method of specifying a plugin by specifying it as a class directly rather than 
    converting a plugin module to a class. So, rather than a plugin module, we could specify:
    ```
    class ImageChecker(OutputCheckersBasePlugin):
        def __call__(self, foo, bar):
            self.compare_outputs()
    ```
4. We could also just stop doing module based plugins and require that all python plugins be written as classes 
    in the first place.

Personally, I prefer option 3 or option 4. 
- Option 3 avoids creating more magic and making module-based plugins more confusing. 
- Option 4 gets rid of all of the magic, but makes implementing a new plugin a little bit more daunting for a non-python programmer and requires lots of changes in lots of places.

I think that option 3 is probably the best compromise solution that avoids the most headaches for everyone including people who have already developed plugins for GeoIPS while option 4 would result in GeoIPS being cleaner and easier to work with.

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [inline geoips/interfaces/module_based/output_checkers.py:154]: Should "test_product" just be another expected method on each plugin within this class?  So rather than passing in the "test_product_func" function, just call self.test_product ?

File: geoips/plugins/modules/procflows/config_based.py
Comment (mindyls) [inline geoips/plugins/modules/procflows/config_based.py:1704]: Ah, so your solution to not having access to the plugin itself within the plugin is to just pass in the plugin explicitly ?  When we decide how to handle this natively within all plugin classes we could presumably remove the "output_checker" argument?

File: geoips/plugins/modules/output_checkers/text_checker.py
Comment (mindyls) [inline geoips/plugins/modules/output_checkers/text_checker.py:122]: Can we remove the need for "test_product_func" ?  Shouldn't it be assumed now that we have separate output checker plugins that the "test_product_func" is just whatever is defined internally to the current output checker?

Do we just need to specify additional required methods for this interface that determine how to compare the outputs, and how to check if a given output is of the type required to use this output checker?

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [inline geoips/interfaces/module_based/output_checkers.py:497]: So this supports calling get_plugin with either the path to a filename or the name of the plugin?  Do we want to be more explicit here, or is this the "pythonic" way of doing it?

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [inline geoips/interfaces/module_based/output_checkers.py:457]: I think we should just require "outputs_match" method on all output checker plugins, then we do not have to have the output checker plugin names listed here explictily.

The only unique part of each of these calls to outputs_match is the prefix to the goodcomps/badcomps strings - perhaps we could just use self.name.upper().  If that happens to break anything downstream with parsing the logs or something, we can fix that (I don't think it will have any impact).

**Review by mindyls — CHANGES_REQUESTED** [review]
Just a couple more organizational changes I think and it should be good.  I think it should be fine to require:
* outputs_match
* correct_type
methods on every output checker plugin?  That is what it means to be an output chekcer, you need to have a way to define that a given file is of the current type, and you need to have a way to define how to compare two files of that type?  So should be fine to just require that?  This would allow us to remove all instances of "test_products_func", which would also allow us to remove the explicit list of plugin names from within the output checkers interface module (in the test_products method).  Thoughts?

File: geoips/interfaces/module_based/output_checkers.py
Comment (evrose54) [reply geoips/interfaces/module_based/output_checkers.py:497]: @jsolbrig recommended I implement this function in this manner. However, I'm happy to leave more comments in this function so it's not as "black-magic" as it is currently.

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [inline geoips/interfaces/module_based/output_checkers.py:95]: @evrose54 Maybe right at the beginning of the "compare_outputs" module just raise an informative error if "outputs_match" or "correct_type" are not defined?  Note this will likely not work for the gzip and gunzip ones (since they are slightly different...), but I *think* it is ok to just handle those explicitly and slightly differently (since they apply to any file format - it's just a way to gunzip before comparing what the type really is, and gzip when you're done).

File: geoips/interfaces/base.py
Comment (jsolbrig) [inline geoips/interfaces/base.py:257]: I think we should move this function onto the BaseModuleInterface class in the same way that plugin_yaml_to_obj is attached to BaseYamlInterface. It should also be a classmethod.

File: geoips/interfaces/base.py
Comment (jsolbrig) [inline geoips/interfaces/base.py:669]: If we move plugin_module_to_obj onto this class as a classmethod, this line can be reverted.

File: geoips/interfaces/base.py
Comment (jsolbrig) [inline geoips/interfaces/base.py:676]: If we move plugin_module_to_obj onto this class as a classmethod, this line can be reverted.

File: geoips/interfaces/module_based/output_checkers.py
Comment (jsolbrig) [reply geoips/interfaces/module_based/output_checkers.py:449]: Yeah, those should be added to required_args here I think.

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [inline geoips/interfaces/module_based/output_checkers.py:221]: Where does rezip get set ?

Perhaps all we want to do here is right at the beginning of this function if the extension is .gz, we gunzip then set rezip=True (so not global), then at the end if rezip is true, we rezip it.  So everything is completely internal to this "compare_outputs" function?

Also, I think what @jsolbrig was suggesting was that we could ALSO gunzip the test output (what is stored in the repo) when we actually do the comparison, then re-zip it - which would prevent the unzipped output sitting in the repo all the time.

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [inline geoips/interfaces/module_based/output_checkers.py:194]: So maybe right here you gunzip each output product.

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [inline geoips/interfaces/module_based/output_checkers.py:437]: I think here at the very end you would have to re-zip all of the files found in "compare_path" as appropriate (so you may need to do some tracking of those in a dictionary or something)

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [inline geoips/interfaces/module_based/output_checkers.py:189]: Probably right here, if we wanted to implement gzip/gunzip for the test products (the ones stored in the repo), you'd have to loop through them all and gunzip any that end in .gz.  Then you'd probably have to have a dictionary of "rezips" to determine which of the files need to be rezipped at the end.  ie rezip[filename] = True or rezip[filename] = False

**Review by mindyls — APPROVED** [review]
Please review and resolve Jeremy's conversations - let me know if any additional issues need to be created resulting from these comments

File: geoips/plugins/modules/output_checkers/geotiff_checker.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/geotiff_checker.py]: If this just needs an issue opened, let's open the issue, reference it here, then resolve this conversations.

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [reply geoips/interfaces/module_based/output_checkers.py]: I'm ok with this, if that's how we want it to work.

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [reply geoips/interfaces/module_based/output_checkers.py:503]: If we can just add an issue for thhis, reference it here then we can resolve the conversation

File: geoips/interfaces/base.py
Comment (evrose54) [reply geoips/interfaces/base.py:676]: I'm going to create an issue for this. I don't want to include too many changes in this PR, and I feel that it would be best to have this separated in another branch. That way, if anything goes wrong, we can track it accordingly.

File: geoips/interfaces/base.py
Comment (evrose54) [reply geoips/interfaces/base.py:257]: This issue will also fix your comments below.

File: geoips/interfaces/base.py
Comment (evrose54) [reply geoips/interfaces/base.py:669]: See https://github.com/NRLMMD-GEOIPS/geoips/pull/351#discussion_r1355580586

File: geoips/interfaces/base.py
Comment (evrose54) [reply geoips/interfaces/base.py:257]: I'll create a separate branch for this. Don't want to include too many changes in this branch, and it will be useful to have this separated just in case something goes wrong and we need to accurately track/source those errors.

Comment (mindyls) [reply geoips/interfaces/base.py:257]: https://github.com/NRLMMD-GEOIPS/geoips/pull/369

File: geoips/interfaces/base.py
Comment (mindyls) [reply geoips/interfaces/base.py:669]: https://github.com/NRLMMD-GEOIPS/geoips/pull/369

File: geoips/interfaces/base.py
Comment (mindyls) [reply geoips/interfaces/base.py:676]: https://github.com/NRLMMD-GEOIPS/geoips/pull/369

File: geoips/interfaces/module_based/output_checkers.py
Comment (mindyls) [reply geoips/interfaces/module_based/output_checkers.py:503]: https://github.com/NRLMMD-GEOIPS/geoips/issues/370
I created an issue.  Resolving.

Comment (mindyls) [reply geoips/interfaces/module_based/output_checkers.py:503]: https://github.com/NRLMMD-GEOIPS/geoips/issues/363
Oops, I had already created an issue...

File: geoips/plugins/modules/output_checkers/geotiff_checker.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/geotiff_checker.py]: This has an issue: https://github.com/NRLMMD-GEOIPS/geoips/issues/362
Resolving.

File: geoips/plugins/modules/output_checkers/utils/compare_outputs.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/utils/compare_outputs.py]: This has been addressed.  Resolving.

File: geoips/plugins/modules/output_checkers/geotiff.py
Comment (mindyls) [inline geoips/plugins/modules/output_checkers/geotiff.py:45]: One more stupid question, then I am good.
@jsolbrig @evrose54 are we good with the "required_methods" names of "correct_type" and "outputs_match" ?

Fine with me, just want to make sure we think about the names before we formalize/finalize them.

**Review by mindyls — APPROVED** [review]
One more comment to resolve, then I am good to go!

File: geoips/plugins/modules/output_checkers/geotiff.py
Comment (evrose54) [reply geoips/plugins/modules/output_checkers/geotiff.py:45]: The only change I would _maybe_ recommend is changing _correct_type_ to _correct_product_type_ or _correct_output_type_. Up to you though. I'm fine with either!

Comment (mindyls) [reply geoips/plugins/modules/output_checkers/geotiff.py:45]: Or "correct_file_type" since it's not like a python type or anything... I'll leave this one up to @jsolbrig , he is particular about names :-)

Comment (mindyls) [reply geoips/plugins/modules/output_checkers/geotiff.py:45]: @evrose54 per @jsolbrig , Update to "correct_file_format", and open an issue to investigate "filetype" module to use rather than exclusively checking file extensions for determining file types.

THEN YOU CAN RESOLVE AND WE CAN MERGE!

**Review by cpcamacho — APPROVED** [review]

**Pull request merged**
